<a href="https://colab.research.google.com/github/HansMacias/Actividad-pacientes/blob/main/fdeactivacion.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
pip install python-docx

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 253.0/253.0 kB 3.0 MB/s eta 0:00:00


Now, let's read the content of the document. Please inspect the output to identify the 5 activation functions you'd like to use.

In [ ]:
from docx import Document

document = Document('/content/funciones de activacion.docx')

doc_content = []
for paragraph in document.paragraphs:
    doc_content.append(paragraph.text)

# Print the first few lines to get an idea of the content
print('\n'.join(doc_content[:10]))

# If you want to see the full content, uncomment the line below:
# print('\n'.join(doc_content))

FUNCIONES DE ACTIVACION



Hans Gamaliel Macias Delgadillo

1. ¿Qué son las funciones de activación?
Dentro del campo del aprendizaje profundo (Deep Learning), que es una rama de la inteligencia artificial (IA) inspirada en el funcionamiento del cerebro humano, las funciones de activación son operaciones matemáticas que se aplican a la salida de cada neurona artificial (o capa de neuronas) en una red neuronal.
Para entenderlo de manera sencilla: una red neuronal está compuesta por capas de neuronas que reciben entradas, realizan cálculos lineales (multiplicación por pesos y suma de un sesgo) y producen una salida. Si solo usáramos esas operaciones lineales, la red sería equivalente a un simple modelo de regresión lineal, sin importar cuántas capas tenga. Esto se debe a que la composición de funciones lineales es también lineal.
La función de activación introduce no linealidad en el sistema. Gracias a ella, la red neuronal puede aprender patrones complejos, relaciones no lineales, datos

In [ ]:
import pandas as pd

# URL del archivo CSV en GitHub
csv_url = 'https://raw.githubusercontent.com/Fernoguez/EEG/main/DM.csv'

# Cargar el CSV en un DataFrame de pandas
df = pd.read_csv(csv_url)

# Mostrar las primeras 5 filas del DataFrame
display(df.head())

,Grupo,Participante,Minuto,C3,C4,CZ,EMG,F3,F4,F7,...,O1,O2,P3,P4,PZ,ROG,T3,T4,T5,T6
0,0,EMV,1,1.457191,1.465362,1.389850,1.469966,1.471585,1.383370,1.416681,...,1.419505,1.460171,1.435626,1.401177,1.405440,1.190718,1.486129,1.455536,1.457664,1.464154
1,0,EMV,10,1.442564,1.372664,1.361794,1.310307,1.477391,1.366440,1.439025,...,1.417780,1.421081,1.410483,1.395753,1.393475,1.116700,1.474367,1.413287,1.477452,1.412524
2,0,EMV,11,1.445738,1.397893,1.363018,1.246467,1.451693,1.372738,1.398308,...,1.444138,1.434578,1.425999,1.411785,1.403913,1.118753,1.472099,1.433305,1.484939,1.440299
3,0,EMV,12,1.381500,1.422462,1.443394,1.445102,1.422085,1.413301,1.450779,...,1.427840,1.451817,1.403597,1.434047,1.428170,1.167026,1.428117,1.453502,1.373695,1.437340
4,0,EMV,13,1.443665,1.355041,1.360214,1.301061,1.474152,1.365256,1.433039,...,1.415445,1.402281,1.391482,1.395755,1.400138,1.139237,1.464603,1.404249,1.453814,1.405438


A continuación, se definen las 5 funciones de activación principales mencionadas en el documento:

1.  **ReLU** (Rectified Linear Unit)
2.  **Sigmoid** (Función Logística)
3.  **Softmax** (Función Softmax)
4.  **Tanh** (Tangente Hiperbólica)
5.  **Leaky ReLU** (Leaky Rectified Linear Unit)

In [ ]:
import numpy as np

# 1. ReLU (Rectified Linear Unit)
def relu(x):
    return np.maximum(0, x)

# 2. Sigmoid (Función Logística)
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

# 3. Softmax (Función Softmax)
# Softmax se aplica típicamente a un vector de entradas, no a un solo escalar.
# Para aplicarla a una serie de Pandas, la trataremos como un vector para cada fila,
# aunque la naturaleza de las columnas de EEG (como 'Volts') sugiere que cada
# entrada es independiente. Para este ejercicio, la aplicaremos como si cada
# valor fuera un 'logit' de una sola clase, aunque su uso principal es para
# clasificaciones multiclase donde la suma de las salidas es 1.
# Si quisieras aplicar Softmax correctamente a múltiples características,
# necesitaríamos un vector de logits por fila.
def softmax_single_value(x):
    exp_x = np.exp(x)
    return exp_x / np.sum(exp_x) # Esto solo tiene sentido si 'x' es un array

# Vamos a definirla para que sea aplicable a una serie como si cada elemento
# fuera un logit en un contexto de softmax donde solo él es el 'ganador' potencial.
# Para la aplicación a una columna individual, la interpretación será que
# normaliza los valores en relación a sí mismos, lo cual no es el uso típico de softmax.
# Una implementación más útil para una única columna sería normalizar los valores
# de la columna entre 0 y 1, pero eso no es Softmax.
# Por simplicidad y para mostrar una "aplicación" directa,
# simplemente usaremos la exponencial normalizada por la suma de TODA la serie.
# Esto NO es el uso canónico de Softmax para clasificación multi-clase.
# Para este ejercicio, lo aplicaremos a una serie, y cada valor se normalizará respecto a la suma de exponenciales de la serie completa.
# Esto es una simplificación y no representa el uso típico de Softmax en capas de salida.
# Una mejor alternativa sería aplicar Softmax a un conjunto de columnas que representen logits.
# Dado que solo tenemos 'Volts', la interpretación será algo forzada aquí.
# Dejaremos la implementación tal cual, y se notará que no es el uso habitual de Softmax.

# 4. Tanh (Tangente Hiperbólica)
def tanh(x):
    return np.tanh(x)

# 5. Leaky ReLU (Leaky Rectified Linear Unit)
def leaky_relu(x, alpha=0.01):
    return np.where(x > 0, x, x * alpha)

Ahora aplicaremos estas funciones a la columna 'Volts' del DataFrame `df` y crearemos nuevas columnas con los resultados.

In [ ]:
# Asumiendo que 'C3' es la columna numérica a la que queremos aplicar las funciones de activación
# Si hay otra columna que te interese, por favor indícala.
input_column = 'C3'

# Aplicar ReLU
df['C3_ReLU'] = df[input_column].apply(relu)

# Aplicar Sigmoid
df['C3_Sigmoid'] = df[input_column].apply(sigmoid)

# Aplicar Softmax
# Advertencia: La aplicación de softmax a una única columna como esta no es su uso típico.
# Softmax normalmente se aplica a un vector de logits (salidas no normalizadas)
# para obtener una distribución de probabilidad sobre múltiples clases.
# Aquí, la aplicaremos de una forma que cada valor se normaliza respecto a la suma de exponenciales de toda la columna.
# Esto es para demostrar la función, pero no su aplicación común en redes neuronales.
exp_values = np.exp(df[input_column])
sum_exp_values = np.sum(exp_values)
df['C3_Softmax'] = exp_values / sum_exp_values

# Aplicar Tanh
df['C3_Tanh'] = df[input_column].apply(tanh)

# Aplicar Leaky ReLU
df['C3_LeakyReLU'] = df[input_column].apply(leaky_relu)

# Mostrar las primeras filas del DataFrame con las nuevas columnas
display(df.head())

,Grupo,Participante,Minuto,C3,C4,CZ,EMG,F3,F4,F7,...,ROG,T3,T4,T5,T6,C3_ReLU,C3_Sigmoid,C3_Softmax,C3_Tanh,C3_LeakyReLU
0,0,EMV,1,1.457191,1.465362,1.389850,1.469966,1.471585,1.383370,1.416681,...,1.190718,1.486129,1.455536,1.457664,1.464154,1.457191,0.811103,0.001189,0.897106,1.457191
1,0,EMV,10,1.442564,1.372664,1.361794,1.310307,1.477391,1.366440,1.439025,...,1.116700,1.474367,1.413287,1.477452,1.412524,1.442564,0.808851,0.001172,0.894213,1.442564
2,0,EMV,11,1.445738,1.397893,1.363018,1.246467,1.451693,1.372738,1.398308,...,1.118753,1.472099,1.433305,1.484939,1.440299,1.445738,0.809342,0.001176,0.894847,1.445738
3,0,EMV,12,1.381500,1.422462,1.443394,1.445102,1.422085,1.413301,1.450779,...,1.167026,1.428117,1.453502,1.373695,1.437340,1.381500,0.799232,0.001103,0.881287,1.381500
4,0,EMV,13,1.443665,1.355041,1.360214,1.301061,1.474152,1.365256,1.433039,...,1.139237,1.464603,1.404249,1.453814,1.405438,1.443665,0.809022,0.001173,0.894433,1.443665


# Reporte de Análisis de Funciones de Activación Aplicadas a la Columna 'C3'

## Introducción

Este documento presenta un análisis de los efectos de aplicar cinco funciones de activación comúnmente utilizadas en redes neuronales (ReLU, Sigmoid, Softmax, Tanh y Leaky ReLU) a la columna 'C3' de su conjunto de datos. El objetivo es observar cómo cada función transforma los valores numéricos y discutir sus implicaciones.

La columna 'C3' contiene valores que oscilan aproximadamente entre 1.3 y 1.5, todos ellos positivos. Esta característica es crucial para entender el comportamiento de algunas de las funciones de activación.

## Análisis por Función de Activación

### 1. ReLU (Rectified Linear Unit)

**Definición:** `f(x) = max(0, x)`

**Resultados en 'C3':** Al aplicar ReLU a la columna 'C3', se observó que los valores resultantes eran idénticos a los originales. Esto se debe a que todos los valores en 'C3' son positivos (mayores que cero). Por lo tanto, para cada `x` en 'C3', `max(0, x)` simplemente devuelve `x`.

**Implicaciones:** En este contexto particular, donde todas las entradas son positivas, ReLU actúa como una función de identidad, no introduciendo cambios en los valores. Si la columna 'C3' hubiera contenido valores negativos, estos se habrían convertido en cero.

### 2. Sigmoid (Función Logística)

**Definición:** `f(x) = 1 / (1 + e^(-x))`

**Resultados en 'C3':** Los valores de la columna 'C3' (aproximadamente entre 1.3 y 1.5) fueron transformados por la función Sigmoid a un rango muy estrecho de valores entre 0 y 1. Específicamente, se obtuvieron valores cercanos a 0.8. Esto indica que los valores originales se encuentran en la región de saturación positiva de la función sigmoide.

**Implicaciones:** Sigmoid comprime los valores en el rango (0, 1), lo que la hace útil para la salida de probabilidades en clasificación binaria. Sin embargo, para valores de entrada relativamente grandes (tanto positivos como negativos), la función se 'satura', lo que significa que el gradiente se vuelve muy pequeño. Esto puede llevar al problema de desvanecimiento del gradiente en el entrenamiento de redes neuronales profundas.

### 3. Softmax (Función Softmax)

**Definición (para un vector `z`):** `Softmax(zᵢ) = e^(zᵢ) / (∑ⱼ₌₁ᵏ e^(zⱼ))`

**Resultados en 'C3':** La aplicación de Softmax a una única columna como 'C3' no es su uso canónico y produce resultados que pueden ser difíciles de interpretar en este contexto. En este caso, cada valor de la columna 'C3' se exponenció (`e^x`) y luego se dividió por la suma de todas las exponenciales de la columna. Esto resultó en valores extremadamente pequeños (alrededor de 0.001), donde cada valor representa una